Go 1.5 support + Refactor

kouno opened the pull request:
I fixed the regexp for Go 1.5 as explained on https://golang.org/pkg/runtime/

> Currently, it is:
> 	gc # @#s #%: #+...+# ms clock, #+...+# ms cpu, #->#-># MB, # MB goal, # P

I ended up also adding tests and fixing a few things:
- It was parsing Stderr and printing on Stdout (via Println). It now always write to Stderr.
- I added a flag to be able to set a fixed port.

davecheney commented:
Thank you, I'll take a look tonight

On Sun, 11 Oct 2015 09:09 Vincent Bonmalais <notifications@github.com>
wrote:

> I fixed the regexp for Go 1.5 as explained on
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/davecheney/gcvis/pull/19
> Commit Summary
>
>    - Update parser to support Golang 1.5
>    - Fix spelling and space
>    - Refactor and test HTTP server and subcommand
>
> File Changes
>
>    - *M* graph.go
>    <https://github.com/davecheney/gcvis/pull/19/files#diff-0> (4)
>    - *A* http_server.go
>    <https://github.com/davecheney/gcvis/pull/19/files#diff-1> (61)
>    - *A* http_server_test.go
>    <https://github.com/davecheney/gcvis/pull/19/files#diff-2> (59)
>    - *M* main.go
>    <https://github.com/davecheney/gcvis/pull/19/files#diff-3> (52)
>    - *M* parser.go
>    <https://github.com/davecheney/gcvis/pull/19/files#diff-4> (54)
>    - *M* parser_test.go
>    <https://github.com/davecheney/gcvis/pull/19/files#diff-5> (54)
>    - *A* subcommand.go
>    <https://github.com/davecheney/gcvis/pull/19/files#diff-6> (51)
>    - *A* subcommand_test.go
>    <https://github.com/davecheney/gcvis/pull/19/files#diff-7> (49)
>    - *D* subprocess.go
>    <https://github.com/davecheney/gcvis/pull/19/files#diff-8> (21)
>
> Patch Links:
>
>    - https://github.com/davecheney/gcvis/pull/19.patch
>    - https://github.com/davecheney/gcvis/pull/19.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/davecheney/gcvis/pull/19>.
>

davecheney commented on subcommand.go:
```diff
@@ -0,0 +1,51 @@
+	PipeRead *os.File
+}
+
+type Stream int
```
drop this please

davecheney commented on subcommand.go:
```diff
@@ -0,0 +1,51 @@
+type Stream int
+
+const (
+	StreamStdout Stream = 1 << iota
```
drop the alias, just make it StreamStdout = 1 << iota, etc

davecheney commented on subcommand.go:
```diff
@@ -0,0 +1,51 @@
+	StreamStderr
+)
+
+func NewSubCommand(args []string, stream Stream) SubCommand {
```
s/Stream/int, this doesn't make this more typesafe, someone can still call this with `42`

davecheney commented on subcommand.go:
```diff
@@ -0,0 +1,51 @@
+
+	switch stream {
+	case StreamStdout:
+		cmd.Stdout = pipeWrite
```
Something is wrong with this logic, if you call a command which fails, and writes to stderr, stderr is lost. At master stderr is forwarded through to the caller.

kouno replied:
I'm not sure I'm understanding this comment. The only difference between master and this branch is the switch statement (and I only added it to make it testable). In `main.go`, I'm still using `StreamStderr`, which means this should behave the same as master.

Speaking of failing commands, I can see now a race condition where `s.cmd.Run()` fails and `log.Fatal(err)` exits before the parser has time to finish its job. But that would apply to the master branch too, wouldn't it?

I will have another run to add tests around this scenario.

davecheney replied:
Thanks. I may not have been clear. What I see is when the command under test runs, any output that gets sent to stderr is lost.,the command I was testing was godoc.



> On 12 Oct 2015, at 19:38, Vincent Bonmalais <notifications@github.com> wrote:
> 
> In subcommand.go:
> 
> > +)
> > +
> > +func NewSubCommand(args []string, stream Stream) SubCommand {
> > +	pipeRead, pipeWrite, err := os.Pipe()
> > +	if err != nil {
> > +		log.Fatal(err)
> > +	}
> > +
> > +	env := append(os.Environ(), "GODEBUG=gctrace=1")
> > +	cmd := exec.Command(args[0], args[1:]...)
> > +	cmd.Env = env
> > +	cmd.Stdin = os.Stdin
> > +
> > +	switch stream {
> > +	case StreamStdout:
> > +		cmd.Stdout = pipeWrite
> I'm not sure I'm understanding this comment. The only difference between master and this branch is the switch statement (and I only added it to make it testable). In main.go, I'm still using StreamStderr, which means this should behave the same as master.
> 
> Speaking of failing commands, I can see now a race condition where s.cmd.Run() fails and log.Fatal(err) exits before the parser has time to finish its job. But that would apply to the master branch too, wouldn't it?
> 
> I will have another run to add tests around this scenario.
> 
> —
> Reply to this email directly or view it on GitHub.
>

davecheney commented:
Thanks for this. A few small nits and one bug where stderr appears to be eaten.

kouno commented:
@davecheney I updated the PR. I covered a few scenarios in the tests such as non-zero exit codes and I fixed a bug where if the subcommand exits early, `gcvis` would stay open.

This was caused by `sc.Scan` waiting for an EOF but never getting it since the pipe was staying open. See: https://github.com/kouno/gcvis/commit/4fc3c17b3667986a344a524b331f9c16361a55a3#diff-363c6f0b9a44ea80c5c84761577d72f0R40.

kouno commented:
Let me know if everything looks good and I'll rebase and clean up the fixup commit.

davecheney commented on subcommand.go:
```diff
@@ -0,0 +1,51 @@
+
+type SubCommand struct {
+	cmd      *exec.Cmd
+	PipeRead *os.File
```
Does this need to be an *os.File, can it be an io.ReadCloser ?

kouno replied:
Indeed it can. I'll update that.

davecheney commented on subcommand.go:
```diff
@@ -0,0 +1,41 @@
+
+type SubCommand struct {
+	cmd       *exec.Cmd
+	PipeRead  *os.File
```
Can these be made io.ReadCloser, io.WriterClosers ?

davecheney commented on subcommand.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+func (s *SubCommand) Run() {
+	if err := s.cmd.Run(); err != nil {
```
s.Err = s.cmd.Run()

is s.Err checked anywhere ? If it Subcommand is run in another goroutine, does there need to be a lock to check s.Err is thread safe ?

kouno replied:
`s.Err` is checked in the main function. However, I don't need a mutex.

There is two ways at the moment to know if `s.Run()` returned an error:

1. Call `Run()` synchronously and check `s.Err`.
2. Use a goroutine to call `Run()` and read the pipe until you reach EOF, and then check `s.Err`.

In the main function, I am in the second scenario as I am waiting for the parser to send `true` through `parser.done` (see https://github.com/davecheney/gcvis/pull/19/files#diff-7ddfb3e035b42cd70649cc33393fe32cR55 and https://github.com/davecheney/gcvis/pull/19/files#diff-f615844d3497ff38db57e459d6ef657bR78). Receiving `parser.done` means that `sc.Scan` reached EOF.

kouno commented:
@davecheney fixed your feedback. I also commented on the Lock. Let me know if that sounds good.

davecheney commented on parser.go:
```diff
@@ -2,54 +2,83 @@ package main
+		p.NoMatchChan <- line
 	}
 
 	if err := sc.Err(); err != nil {
```
p.Err = sc.Err() 

if sc.Err() returns nil, then p.Err will be nil, there is no need to check explicitly

davecheney commented on http_server_test.go:
```diff
@@ -0,0 +1,59 @@
+		t.Errorf("Error while reading response body: %v", err)
+	}
+
+	w := &bytes.Buffer{}
```
This can be written as

var w bytes.Buffer, and &w below if you prefer

davecheney commented on parser.go:
```diff
@@ -2,54 +2,83 @@ package main
+		p.Err = err
 	}
+
+	p.done <- true
```
close(p.done), then you don't need to make it a buffered channel

davecheney commented on subcommand_test.go:
```diff
@@ -0,0 +1,67 @@
+			t.Errorf("line is not equal to 'hello world': '%v'", string(content))
+		}
+
+		done <- true
```
close(done) // as above

davecheney commented on main.go:
```diff
@@ -41,41 +31,39 @@ func main() {
+		case output := <-parser.NoMatchChan:
+			fmt.Fprintln(os.Stderr, output)
+		case <-parser.done:
+			if parser.Err != nil {
```
this is not a data race because the send on parser.done happens before parser.Err

davecheney commented on main.go:
```diff
@@ -41,41 +31,39 @@ func main() {
+				os.Exit(1)
+			}
+
+			if subcommand.Err != nil {
```
This is a race as there is no synchronisation between the goroutine setting .Err and this one reading it.

A simple solution might be to have subcommnd.Run just print the error, and drop the os.Exit(1) case here.

davecheney commented on parser.go:
```diff
@@ -2,54 +2,83 @@ package main
 
+	if p.gcRegexp == nil {
+		// Set regexp based on Golang version
+		if runtime.Version() == "go1.5" {
```
This will fail if the runtime version is not go1.5, for example
```
lucky(~/src) % ~/go1.5/bin/go run version.go
go1.5.1
```
You could use a conditional compilation, ie break it out into two files with build tags // +build !go1.4, etc, or use strings.HasPrefix

kouno replied:
I decided to parse the output of both 1.4 and 1.5. We can't guarantee which version of Go the binary was created with, so I decided to go with this solution.

davecheney commented:
LGTM with a few more minor cleanups.

davecheney commented:
Btw. If you want to go ahead and we can land this now, then tackle the rest in followups.

davecheney commented:
Above, when I said "we", I was planning on adding you as a committer, but it looks like I already did before. So, when you're happy with this PR, please merge it and we can fix any remaining issues in followups.

davecheney commented:
Thanks!

On Sat, 17 Oct 2015, 09:22 Vincent Bonmalais <notifications@github.com>
wrote:

> Merged #19 <https://github.com/davecheney/gcvis/pull/19>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/davecheney/gcvis/pull/19#event-438025297>.
>
